src/ext/java/org/opentripplanner/ext/fares/impl/DefaultFareService.java

```diff
@@ -102,18 +104,89 @@ public ItineraryFares calculateFares(Itinerary itinerary) {
+      .stream()
+      .collect(
+        Collectors.toMap(
+          Map.Entry::getKey,
+          rules ->
+            rules
+              .getValue()
+              .stream()
+              .collect(Collectors.groupingBy(rule -> rule.getFareAttribute().getId().getFeedId()))
+        )
+      );
 
     ItineraryFares fare = ItineraryFares.empty();
     boolean hasFare = false;
-    for (Map.Entry<FareType, Collection<FareRuleSet>> kv : fareRulesPerType.entrySet()) {
-      FareType fareType = kv.getKey();
-      Collection<FareRuleSet> fareRules = kv.getValue();
-      // Get the currency from the first fareAttribute, assuming that all tickets use the same currency.
-      if (fareRules.size() > 0) {
-        Currency currency = Currency.getInstance(
-          fareRules.iterator().next().getFareAttribute().getCurrencyType()
+    for (FareType fareType : fareRulesPerType.keySet()) {
+      List<FareComponent> components = new ArrayList<>();
+      List<Money> fares = new ArrayList<>();
+      ItineraryFares currentFare = ItineraryFares.empty();
+      boolean legWithoutRulesFound = false;
+      boolean legsWithoutMatchingRulesFound = false;
+      boolean fareTypeHasFare = false;
+      for (String feedId : fareLegsByFeed.keySet()) {
+        var fareRules = fareRulesByTypeAndFeed.get(fareType).get(feedId);
+
+        // Get the currency from the first fareAttribute, assuming that all tickets use the same currency.
+        if (fareRules != null && fareRules.size() > 0) {
+          Currency currency = Currency.getInstance(
+            fareRules.iterator().next().getFareAttribute().getCurrencyType()
+          );
+          boolean feedHasFare = populateFare(
+            currentFare,
+            currency,
+            fareType,
+            fareLegsByFeed.get(feedId),
+            fareRules
+          );
+
+          if (!feedHasFare) {
+            legsWithoutMatchingRulesFound = true;
+          }
+          hasFare = feedHasFare || hasFare; // Other feeds might still have fare for some legs
+
+          components.addAll(currentFare.getComponents(fareType));
+          fare.addFare(fareType, currentFare.getFare(fareType));
+          fares.add(currentFare.getFare(fareType));
+
+          // If all the legs are from one feed, consider itinerary products
+          if (fareLegs.equals(fareLegsByFeed.get(feedId))) {
+            fare.addItineraryProducts(currentFare.getItineraryProducts());
+          }
+        } else {
+          legWithoutRulesFound = true;
+        }
+      }
+
+      fare.addFareComponent(fareType, components);
+
+      // No fares will be discovered after this point
+      if (!hasFare) {
+        continue;
+      }
+
+      // Accumulate the final price of the fare or indicate that no final fare could be found
+      if (legWithoutRulesFound || legsWithoutMatchingRulesFound) {
+        fare.addFare(
+          fareType,
+          Money.ofFractionalAmount(fares.get(0).currency(), UNKNOWN_FARE_PRICE)
```

Comment:
Actually, I'm reconsidering this. If no fare could be found, should we simply skip this? Does _your_ UI expect there to be a -1 cent?

Reply:
Yes, it specifically checks if the price is -1 cent and if it is, it warns the user of missing fares if the price. Otherwise no warning is shown.